### PR TITLE
[UR] Add adapter leak-checking tests

### PR DESCRIPTION
**Pull request by kswiecicki**

### CODE REVIEW

**Review by pbalcer — COMMENTED** [review]
can you add a test?

**Comment by kswiecicki** [comment]
> can you add a test?

Done.
Apparently there's a limit to the number of groups in a regex that ctest can compile (it fails as if the match failed) so I've changed `(.*)` to `.*` everywhere.

**Review by pbalcer — APPROVED** [review]

**Comment by pbalcer** [comment]
> > 
> 
> Done. Apparently there's a limit to the number of groups in a regex that ctest can compile (it fails as if the match failed) so I've changed `(.*)` to `.*` everywhere.

Use the python match scripts in the future.

**Comment by kswiecicki** [comment]
> > > 
> > 
> > 
> > Done. Apparently there's a limit to the number of groups in a regex that ctest can compile (it fails as if the match failed) so I've changed `(.*)` to `.*` everywhere.
> 
> Use the python match scripts in the future.

I've looked into the `.match.py` but it seems to support only line-by-line matching. In my case I needed to match some unknown number of backtrace lines.

File: cmake/match.py
Comment (pbalcer) [reply cmake/match.py:84]: no, let's not bump required python version unnecessarily.

Comment (kswiecicki) [reply cmake/match.py]: I've removed the walrus operator.

**Review by PatKamin — APPROVED** [review]

File: cmake/match.py
Comment (pbalcer) [inline cmake/match.py:34]: brief?:P

File: cmake/match.py
Comment (pbalcer) [inline cmake/match.py:85]: this doesn't seem necessary, this condition will never be false as far as I can tell. You can just do a while True: loop and check status at the end (and exit appropriately).

File: cmake/match.py
Comment (pbalcer) [inline cmake/match.py:106]: doesn't seem like there can be more than one active tag?

**Comment by pbalcer** [comment]
I don't think we can merge this is as, it will lead to conflicts with the adapters branch. This patch needs to be rebased on adapters or wait until the branches are merged.

File: cmake/match.py
Comment (kswiecicki) [reply cmake/match.py:35]: Done.

File: cmake/match.py
Comment (kswiecicki) [reply cmake/match.py:105]: In the case below, there would be a point when ignore tag (active until the next match) is applied, then the optional tag (active until the next line) is applied next.
```
{{IGNORE}}
{{OPT}} optional line
```

File: cmake/match.py
Comment (kswiecicki) [reply cmake/match.py]: I've moved the status variable inside the loop and changed the loop condition.
If I were to move the status check at the end, then I would need to check if either input or match file is empty to avoid index error. Logger has such test where both input and match files are empty.

**Comment by kswiecicki** [comment]
> I don't think we can merge this is as, it will lead to conflicts with the adapters branch. This patch needs to be rebased on adapters or wait until the branches are merged.

Alright, I can wait for the adapters branch to be merged with main.

File: cmake/match.py
Comment (kswiecicki) [reply cmake/match.py]: Also, the match case statement (switch equivalent) is not available in our current supported python version.

**Comment by pbalcer** [comment]
this does not appear to be changing anything but tests?

**Comment by kbenzie** [comment]
I'm getting similar failures locally which are the result of missing device binaries although not exactly the same. I'll continue investigating.

**Comment by kbenzie** [comment]
This appears to have been caused by [this change](https://github.com/oneapi-src/unified-runtime/commit/5dc6e464c83fa41045d4f352a300db9b16400d1d#diff-6bc28da4be765fa3d871baab7e4d458635e40dc2b131a573167f24c7cbc7c9b2R217) where `urProgramCreateWithIL` would formerly pass through to `urProgramCreateWithBinary` but no longer does, instead calling `detail::ur::die()`.

What I don't currently understand is why this isn't affecting all PR's.

**Comment by kswiecicki** [comment]
> This appears to have been caused by [this change](https://github.com/oneapi-src/unified-runtime/commit/5dc6e464c83fa41045d4f352a300db9b16400d1d#diff-6bc28da4be765fa3d871baab7e4d458635e40dc2b131a573167f24c7cbc7c9b2R217) where `urProgramCreateWithIL` would formerly pass through to `urProgramCreateWithBinary` but no longer does, instead calling `detail::ur::die()`.
> 
> What I don't currently understand is why this isn't affecting all PR's.

Thank you for the investigation @kbenzie. With @pbalcer help, we were able to figure out the cause of those test fails.
Previous match script had a bug that allowed a test to pass despite not matching the `Aborted` input if all of the match file lines were marked as `{{OPT}}`. As a temporary solution I've added `Aborted` as an expected input, I've seen other adapter tests do that as well.

**Review by pbalcer — APPROVED** [review]

**Review by PatKamin — APPROVED** [review]

**Review by kbenzie — APPROVED** [review]

**Pull request merged**
